homeassistant/components/history/__init__.py

```diff
@@ -300,13 +362,409 @@ def _sorted_significant_states_json(
 
 
 def _entities_may_have_state_changes_after(
-    hass: HomeAssistant, entity_ids: Iterable, start_time: dt
+    hass: HomeAssistant, entity_ids: Iterable, start_time: dt, no_attributes: bool
 ) -> bool:
     """Check the state machine to see if entities have changed since start time."""
     for entity_id in entity_ids:
         state = hass.states.get(entity_id)
+        if state is None:
+            return True
 
-        if state is None or state.last_changed > start_time:
+        state_time = state.last_changed if no_attributes else state.last_updated
+        if state_time > start_time:
             return True
 
     return False
+
+
+def _generate_stream_message(
+    states: MutableMapping[str, list[dict[str, Any]]],
+    start_day: dt,
+    end_day: dt,
+) -> dict[str, Any]:
+    """Generate a history stream message response."""
+    return {
+        "states": states,
+        "start_time": dt_util.utc_to_timestamp(start_day),
+        "end_time": dt_util.utc_to_timestamp(end_day),
+    }
+
+
+@callback
+def _async_send_empty_response(
+    connection: ActiveConnection, msg_id: int, start_time: dt, end_time: dt | None
+) -> None:
+    """Send an empty response when we know all results are filtered away."""
+    connection.send_result(msg_id)
+    stream_end_time = end_time or dt_util.utcnow()
+    _async_send_response(connection, msg_id, start_time, stream_end_time, {})
+
+
+@callback
+def _async_send_response(
+    connection: ActiveConnection,
+    msg_id: int,
+    start_time: dt,
+    end_time: dt,
+    states: MutableMapping[str, list[dict[str, Any]]],
+) -> None:
+    """Send a response."""
+    empty_stream_message = _generate_stream_message(states, start_time, end_time)
+    empty_response = messages.event_message(msg_id, empty_stream_message)
+    connection.send_message(JSON_DUMP(empty_response))
+
+
+async def _async_send_historical_states(
+    hass: HomeAssistant,
+    connection: ActiveConnection,
+    msg_id: int,
+    start_time: dt,
+    end_time: dt,
+    entity_ids: list[str] | None,
+    filters: Filters | None,
+    include_start_time_state: bool,
+    significant_changes_only: bool,
+    minimal_response: bool,
+    no_attributes: bool,
+    send_empty: bool,
+) -> dt | None:
+    """Fetch history significant_states and send them to the client."""
+    states = cast(
```

Comment:
cast here because `get_significant_states` can return non-compressed and compressed but we always use compressed here